src/drivers/ihda.cpp

```diff
@@ -0,0 +1,309 @@
+
+    if (dev.msi_cap()) {
+      this->m_irq = dev.get_legacy_irq();
+    }
+    Events::get().subscribe(this->m_irq, {this, &IHDA::event_handler});
+  }
+  else {
+    assert(0 && "Legacy INT#x not supported");
+  }
+  INFO2("|  |- IRQ: %u", this->m_irq);
+  // MMIO BAR
+  this->m_hdbar = dev.get_bar(0).start;
+  INFO2("|  |- MMIO address: %p", (void*) this->m_hdbar);
+
+  const uint8_t major = read<uint8_t>(0x3);
+  const uint8_t minor = read<uint8_t>(0x2);
+  INFO2("|  |- Version: %d.%d", major, minor);
+
+  this->reset();
+  this->detect_codecs();
+  INFO("IntelHDA", "|  +--[ Codecs: %zu ]", m_codecs.size());
+  m_corb.reg = 0x40;
+  m_rirb.reg = 0x50;
+  this->setup_ring(m_corb, true);
+  this->setup_ring(m_rirb, false);
+
+  for (auto& codec : m_codecs)
+  {
+    // get subnode count
+    auto req = this->get(0, codec.idx, 0x4);
+    codec.start_node = (req.resp >> 16) & 0xFF;
+    const int afg_count = req.resp & 0xff;
+
+    for (int i = 0; i < afg_count; i++)
+    {
+      const int subnode = codec.start_node + i;
+      // get subnode count
+      auto req = this->get(subnode, codec.idx, 0x4);
+      // get function group type
+      auto func_group = this->get(subnode, codec.idx, 0x5);
+
+      codec.afgs.emplace_back();
+      auto& afg = codec.afgs.back();
+      afg.node_id = subnode;
+      afg.type    = func_group.resp;
+      afg.start_node = (req.resp >> 16) & 0xff;
+      afg.node_count = req.resp & 0xff;
+
+      INFO2("|     o--[ Function Group %u ]", i);
+
+      for (uint32_t w = 0; w < afg.node_count; w++)
+      {
+        afg.widgets.emplace_back();
+        auto& widget = afg.widgets.back();
+        widget.m_node_id = afg.start_node + w;
+        // get function group type
+        auto func_group = this->get(widget.nid(), codec.idx, 0x5);
+        // get widget capabilities
+        auto req_caps = this->get(widget.nid(), codec.idx, 0x9);
+        widget.m_caps = req_caps.resp;
+        // get conn list length
+        auto req_cll = this->get(widget.nid(), codec.idx, 0xE);
+        widget.m_conn_listen_len = req_cll.resp;
+        // get conn list entries
+        auto req_fcle = this->cmd(widget.nid(), codec.idx, 0xF02 << 8);
+        widget.m_first_conn_list_entry = req_fcle.resp;
+
+        INFO2("|     +-- Widget %u: %s (%#x)", w, widget.name(), widget.type());
+      }
+      INFO2("|");
+    }
+  }
+
+  // enable interrupts
+
+}
+
+void IHDA::reset()
+{
+  // reset the device and wait
+  this->write<uint32_t>(REG_GCTL, 0);
+  for(int n = 0; n < 6000; n++) asm("pause");
+  // start the device
+  this->write<uint32_t>(REG_GCTL, 1);
+  // wait for bit to reset
+  while ((this->read<uint32_t>(REG_GCTL) & 0x1) == 0) asm("pause");
+}
+
+void IHDA::detect_codecs()
+{
+  uint8_t stat = this->read<uint8_t>(REG_STATESTS);
+  for (int i = 0; i < 16; i++) {
+    if (stat & (1 << i)) {
+      m_codecs.emplace_back();
+      m_codecs.back().idx = i;
+    }
+  }
+}
+
+void IHDA::setup_ring(corb_t& ring, bool corb)
```

Comment:
Any point thinking about memory barriers ?

Reply:
I have tried to use them, especially one just before a branch that checks an MMIO variable, but it made no difference. Im not using those rings atm because the "immediate command" interface works better